test/core/call/request_buffer_test.cc

```diff
@@ -0,0 +1,573 @@
+    ASSERT_THAT(poll_msg, IsReady());
+    ASSERT_TRUE(poll_msg.value().ok());
+    ASSERT_TRUE(poll_msg.value().value().has_value());
+    EXPECT_THAT(poll_msg.value().value().value(), IsTestMessage(i));
+  }
+}
+
+TEST(RequestBufferTest, PushSomeSwitchThenPushPullMessages) {
+  RequestBuffer buffer;
+  EXPECT_EQ(buffer.PushClientInitialMetadata(TestMetadata()), Success{});
+  RequestBuffer::Reader reader(&buffer);
+  auto pull_md = reader.PullClientInitialMetadata();
+  EXPECT_THAT(pull_md(), IsReady());  // value tested elsewhere
+  for (int i = 0; i < 10; i++) {
+    auto pusher = buffer.PushMessage(TestMessage(i));
+    EXPECT_THAT(pusher(), IsReady(40 + 9 * (i + 1)));
+  }
+  buffer.SwitchToStreaming(&reader);
+  for (int i = 0; i < 10; i++) {
+    auto pull_msg = reader.PullMessage();
+    auto poll_msg = pull_msg();
+    ASSERT_THAT(poll_msg, IsReady());
+    ASSERT_TRUE(poll_msg.value().ok());
+    ASSERT_TRUE(poll_msg.value().value().has_value());
+    EXPECT_THAT(poll_msg.value().value().value(), IsTestMessage(i));
+  }
+  for (int i = 0; i < 10; i++) {
+    auto pusher = buffer.PushMessage(TestMessage(i));
+    EXPECT_THAT(pusher(), IsReady(0));
+    auto pull_msg = reader.PullMessage();
+    auto poll_msg = pull_msg();
+    ASSERT_THAT(poll_msg, IsReady());
+    ASSERT_TRUE(poll_msg.value().ok());
+    ASSERT_TRUE(poll_msg.value().value().has_value());
+    EXPECT_THAT(poll_msg.value().value().value(), IsTestMessage(i));
+  }
+}
+
+TEST(RequestBufferTest, HedgeReadMetadata) {
+  RequestBuffer buffer;
+  EXPECT_EQ(buffer.PushClientInitialMetadata(TestMetadata()), Success{});
+  RequestBuffer::Reader reader1(&buffer);
+  RequestBuffer::Reader reader2(&buffer);
+  auto pull_md1 = reader1.PullClientInitialMetadata();
+  auto pull_md2 = reader2.PullClientInitialMetadata();
+  auto poll_md1 = pull_md1();
+  auto poll_md2 = pull_md2();
+  ASSERT_THAT(poll_md1, IsReady());
+  ASSERT_THAT(poll_md2, IsReady());
+  auto value1 = std::move(poll_md1.value());
+  auto value2 = std::move(poll_md2.value());
+  ASSERT_TRUE(value1.ok());
+  ASSERT_TRUE(value2.ok());
+  EXPECT_THAT(*value1, IsTestMetadata());
+  EXPECT_THAT(*value2, IsTestMetadata());
+}
+
+TEST(RequestBufferTest, HedgeReadMetadataSwitchBeforeFirstRead) {
+  RequestBuffer buffer;
+  EXPECT_EQ(buffer.PushClientInitialMetadata(TestMetadata()), Success{});
+  RequestBuffer::Reader reader1(&buffer);
+  buffer.SwitchToStreaming(&reader1);
+  RequestBuffer::Reader reader2(&buffer);
+  auto pull_md1 = reader1.PullClientInitialMetadata();
+  auto pull_md2 = reader2.PullClientInitialMetadata();
+  auto poll_md1 = pull_md1();
+  auto poll_md2 = pull_md2();
+  ASSERT_THAT(poll_md1, IsReady());
+  ASSERT_THAT(poll_md2, IsReady());
+  auto value1 = std::move(poll_md1.value());
+  auto value2 = std::move(poll_md2.value());
+  ASSERT_TRUE(value1.ok());
+  EXPECT_FALSE(value2.ok());
+  EXPECT_THAT(*value1, IsTestMetadata());
+}
+
+TEST(RequestBufferTest, HedgeReadMetadataLate) {
```

Comment:
Does this test actually cover some case that the previous one doesn't?  It's not obvious to me that it exercises a different code path.

Reply:
It's a canary against possible optimizations that treat read ordering differently - right now I think it's not showing up different code paths.